File: app/fsharp-2/wpf/tests/PuyoPuyo.Tests/Domain/GameLogicTests.fs
namespace PuyoPuyo.Tests.Domain

open Xunit
open FsUnit.Xunit
open PuyoPuyo.Domain

module GameLogicTests =

    [<Fact>]
    let ``ぷよペアを左に移動できる`` () =
        // Arrange
        let board = Board.create 6 13
        let pair = PuyoPair.create 3 5 Red Green 0

        // Act
        let result = GameLogic.tryMovePuyoPair board pair Left

        // Assert
        match result with
        | Some movedPair ->
            movedPair.X |> should equal 2
            movedPair.Y |> should equal 5
        | None -> failwith "移動できるはずです"

    [<Fact>]
    let ``ぷよペアを右に移動できる`` () =
        // Arrange
        let board = Board.create 6 13
        let pair = PuyoPair.create 2 5 Red Green 0

        // Act
        let result = GameLogic.tryMovePuyoPair board pair Right

        // Assert
        match result with
        | Some movedPair ->
            movedPair.X |> should equal 3
            movedPair.Y |> should equal 5
        | None -> failwith "移動できるはずです"

    [<Fact>]
    let ``左端では左に移動できない`` () =
        // Arrange
        let board = Board.create 6 13
        let pair = PuyoPair.create 0 5 Red Green 0

        // Act
        let result = GameLogic.tryMovePuyoPair board pair Left

        // Assert
        result |> should equal None

    [<Fact>]
    let ``右端では右に移動できない`` () =
        // Arrange
        let board = Board.create 6 13
        let pair = PuyoPair.create 5 5 Red Green 0

        // Act
        let result = GameLogic.tryMovePuyoPair board pair Right

        // Assert
        result |> should equal None

    [<Fact>]
    let ``回転を考慮して右端で右に移動できない`` () =
        // Arrange
        let board = Board.create 6 13
        // 回転=1(右向き)の場合、2つ目のぷよは右に配置される
        let pair = PuyoPair.create 4 5 Red Green 1

        // Act
        let result = GameLogic.tryMovePuyoPair board pair Right

        // Assert
        result |> should equal None

    [<Fact>]
    let ``通常の回転ができる`` () =
        // Arrange
        let board = Board.create 6 13
        let pair = PuyoPair.create 3 5 Red Green 0

        // Act
        let result = GameLogic.tryRotatePuyoPair board pair

        // Assert
        match result with
        | Some rotated -> rotated.Rotation |> should equal 1
        | None -> failwith "回転できるはずです"

    [<Fact>]
    let ``右端で回転すると左にキックされる`` () =
        // Arrange
        let board = Board.create 6 13
        let pair = PuyoPair.create 5 5 Red Green 0 // 右端、回転状態0（上）

        // Act
        let result = GameLogic.tryRotatePuyoPair board pair

        // Assert
        match result with
        | Some rotated ->
            rotated.Rotation |> should equal 1 // 回転成功
            rotated.X |> should equal 4 // 左に1マスキック
        | None -> failwith "回転できるはずです"

    [<Fact>]
    let ``左端で回転状態2から回転すると右にキックされる`` () =
        // Arrange
        let board = Board.create 6 13
        let pair = PuyoPair.create 0 5 Red Green 2 // 左端、回転状態2（下）

        // Act
        let result = GameLogic.tryRotatePuyoPair board pair

        // Assert
        match result with
        | Some rotated ->
            // 回転状態2→3では2つ目のぷよが左にはみ出すため、右に1マスキック
            rotated.Rotation |> should equal 3 // 回転成功（2→3）
            rotated.X |> should equal 1 // 右に1マスキック
        | None -> failwith "回転できるはずです"

    [<Fact>]
    let ``壁キックできない場合は回転しない`` () =
        // Arrange
        let board = Board.create 6 13
        // 右隣にぷよを配置（壁キックできない状況を作る）
        let board = Board.setCell 4 5 (Filled Blue) board
        let pair = PuyoPair.create 5 5 Red Green 0

        // Act
        let result = GameLogic.tryRotatePuyoPair board pair

        // Assert
        result |> should equal None

File: app/csharp-2/maui/PuyoPuyoMAUI.Tests/Domain/GameLogicTests.cs
using FluentAssertions;
using PuyoPuyoMAUI.Domain;
using Xunit;

namespace PuyoPuyoMAUI.Tests.Domain;

public class GameLogicTests
{
    [Fact]
    public void ぷよペアを左に移動できる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(3, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.TryMovePuyoPair(board, pair, Direction.Left);

        // Assert
        result.Should().NotBeNull();
        result!.X.Should().Be(2);
        result.Y.Should().Be(5);
    }

    [Fact]
    public void ぷよペアを右に移動できる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(2, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.TryMovePuyoPair(board, pair, Direction.Right);

        // Assert
        result.Should().NotBeNull();
        result!.X.Should().Be(3);
        result.Y.Should().Be(5);
    }

    [Fact]
    public void 左端では左に移動できない()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(0, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.TryMovePuyoPair(board, pair, Direction.Left);

        // Assert
        result.Should().BeNull();
    }

    [Fact]
    public void 右端では右に移動できない()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(5, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.TryMovePuyoPair(board, pair, Direction.Right);

        // Assert
        result.Should().BeNull();
    }

    [Fact]
    public void 回転状態1_右向き_のとき右端で右に移動できない()
    {
        // Arrange
        var board = Board.Create(6, 13);
        // 回転状態1のとき、2つ目のぷよは右にある
        var pair = new PuyoPair(4, 5, PuyoColor.Red, PuyoColor.Green, 1);

        // Act
        var result = GameLogic.TryMovePuyoPair(board, pair, Direction.Right);

        // Assert
        result.Should().BeNull();
    }

    [Fact]
    public void ぷよペアが配置可能な位置かチェックできる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(2, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.CanPlacePuyoPair(board, pair);

        // Assert
        result.Should().BeTrue();
    }

    [Fact]
    public void ボードの外は配置不可能()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(-1, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.CanPlacePuyoPair(board, pair);

        // Assert
        result.Should().BeFalse();
    }

    [Fact]
    public void 他のぷよがある場所は配置不可能()
    {
        // Arrange
        var board = Board.Create(6, 13)
            .SetCell(2, 5, Cell.Filled(PuyoColor.Blue));
        var pair = new PuyoPair(2, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.CanPlacePuyoPair(board, pair);

        // Assert
        result.Should().BeFalse();
    }

    [Fact]
    public void 右端で回転すると左にキックされる()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(5, 5, PuyoColor.Red, PuyoColor.Green, 0);  // 右端、回転状態0（上）

        // Act
        var result = GameLogic.TryRotatePuyoPair(board, pair);

        // Assert
        result.Should().NotBeNull();
        result!.Rotation.Should().Be(1);  // 回転成功
        result.X.Should().Be(4);  // 左に1マスキック
    }

    [Fact]
    public void 左端で左向きから回転すると右にキックされる()
    {
        // Arrange
        var board = Board.Create(6, 13)
            .SetCell(0, 4, Cell.Filled(PuyoColor.Blue));  // 上に障害物
        var pair = new PuyoPair(0, 5, PuyoColor.Red, PuyoColor.Green, 3);  // 左端、回転状態3（左）

        // Act
        var result = GameLogic.TryRotatePuyoPair(board, pair);

        // Assert
        result.Should().NotBeNull();
        result!.Rotation.Should().Be(0);  // 回転成功
        result.X.Should().Be(1);  // 右に1マスキック
    }

    [Fact]
    public void 壁キックできない場合は回転しない()
    {
        // Arrange
        var board = Board.Create(6, 13);
        // 右端にぷよを配置（壁キックできない状況を作る）
        var board2 = board.SetCell(4, 5, Cell.Filled(PuyoColor.Blue));
        var pair = new PuyoPair(5, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.TryRotatePuyoPair(board2, pair);

        // Assert
        result.Should().BeNull();
    }

    [Fact]
    public void 通常の回転が可能な場合は壁キックしない()
    {
        // Arrange
        var board = Board.Create(6, 13);
        var pair = new PuyoPair(3, 5, PuyoColor.Red, PuyoColor.Green, 0);

        // Act
        var result = GameLogic.TryRotatePuyoPair(board, pair);

        // Assert
        result.Should().NotBeNull();
        result!.Rotation.Should().Be(1);
        result.X.Should().Be(3);  // 位置は変わらない
    }
}
